trl/experimental/gold/gold_trainer.py

```diff
@@ -0,0 +1,2038 @@
+        Returns:
+            JSD loss for matched tokens
+        """
+        # Reshape to [batch_size * seq_len, vocab_size] format expected by generalized_jsd_loss
+        batch_seq_len, num_matched = student_logits.shape
+
+        student_logits_reshaped = student_logits.view(-1, num_matched)
+        teacher_logits_reshaped = teacher_logits.view(-1, num_matched)
+
+        # Use the GOLD generalized JSD loss implementation that accepts probability inputs
+        jsd_loss = GOLDTrainer.generalized_jsd_loss(
+            student_logits_reshaped,
+            teacher_logits_reshaped,
+            labels=None,  # No masking needed for matched tokens
+            beta=self.beta,  # Standard JSD beta
+            temperature=1.0,  # Already applied in main computation
+            reduction="batchmean",
+            logits_are_probs=True,
+        )
+
+        return jsd_loss
+
+    def _get_start_and_size_answers(self, answer_tensors):
+        answers_index = []
+        answers_size = []
+
+        for answer in answer_tensors:
+            is_value = answer.eq(self.ignore_index)
+            answers_size.append(len(answer) - int(is_value.sum()))
+            indices = is_value.nonzero(as_tuple=True)[0]
+            if len(indices) == 0 or indices[0] != 0:
+                answers_index.append(0)
+            else:
+                diff_indices = indices[1:] - indices[:-1]
+                break_index = (diff_indices != 1).nonzero()
+                length = (break_index[0].item() + 1) if len(break_index) > 0 else len(indices)
+                answers_index.append(length - 1)
+        return answers_index, answers_size
+
+
+class GOLDStudentVLLMSyncCallback(TrainerCallback):
+    """
+    Callback to sync student model weights to vLLM after training steps. This ensures weight syncing happens when
+    DeepSpeed is in a stable state.
+    """
+
+    def __init__(self, trainer):
+        self.trainer = trainer
+
+    def on_step_end(self, args, state: TrainerState, control: TrainerControl, **kwargs):
+        """Sync weights after training step when DeepSpeed is stable."""
+        if (
+            self.trainer.student_use_vllm
+            and state.global_step != self.trainer._last_student_sync_step
+            and state.global_step % self.trainer.student_vllm_sync_frequency == 0
+        ):
+            # Check if this is a step where gradients are synchronized
+            # This happens at the end of gradient accumulation cycles
+            if hasattr(self.trainer.accelerator, "sync_gradients") and self.trainer.accelerator.sync_gradients:
+                self.trainer._move_student_model_to_vllm()
+                self.trainer._last_student_sync_step = state.global_step
+
+
+class GOLDTrainer(SFTTrainer):
+    _tag_names = ["trl", "gold"]
+
+    def __init__(
+        self,
+        model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
+        model_revision: Optional[str] = None,
```

Comment:
any reason for having model_revision in the trainer arg instead of trainer config?